test/cpp/naming/address_sorting_test.cc

```diff
@@ -0,0 +1,540 @@
+  char* host;
+  char* port;
+  grpc_resolved_address resolved_addr;
+  gpr_split_host_port(test_addr.dest_addr.c_str(), &host, &port);
+  if (test_addr.family == AF_INET) {
+    sockaddr_in in_dest;
+    in_dest.sin_port = htons(atoi(port));
+    ;
+    in_dest.sin_family = AF_INET;
+    int zeros[2]{0, 0};
+    memcpy(&in_dest.sin_zero, &zeros, 8);
+    GPR_ASSERT(inet_pton(AF_INET, host, &in_dest.sin_addr) == 1);
+    memcpy(&resolved_addr.addr, &in_dest, sizeof(sockaddr_in));
+    resolved_addr.len = sizeof(sockaddr_in);
+  } else {
+    GPR_ASSERT(test_addr.family == AF_INET6);
+    sockaddr_in6 in6_dest;
+    in6_dest.sin6_port = htons(atoi(port));
+    in6_dest.sin6_family = AF_INET6;
+    in6_dest.sin6_flowinfo = 0;
+    in6_dest.sin6_scope_id = 0;
+    GPR_ASSERT(inet_pton(AF_INET6, host, &in6_dest.sin6_addr) == 1);
+    memcpy(&resolved_addr.addr, &in6_dest, sizeof(sockaddr_in6));
+    resolved_addr.len = sizeof(sockaddr_in6);
+  }
+  return resolved_addr;
+}
+
+class MockSocketFactory : public address_sorting_socket_factory {
+ public:
+  MockSocketFactory(bool ipv4_supported, bool ipv6_supported,
+                    std::map<std::string, TestAddress> dest_addr_to_src_addr)
+      : ipv4_supported_(ipv4_supported),
+        ipv6_supported_(ipv6_supported),
+        dest_addr_to_src_addr_(dest_addr_to_src_addr),
+        fd_to_getsockname_return_vals_(std::map<int, TestAddress>()),
+        cur_socket_(0) {}
+
+  int Socket(int domain, int type, int protocol) {
+    EXPECT_TRUE(domain == AF_INET || domain == AF_INET6);
+    if ((domain == AF_INET && !ipv4_supported_) ||
+        (domain == AF_INET6 && !ipv6_supported_)) {
+      errno = EAFNOSUPPORT;
+      return -1;
+    }
+    return cur_socket_++;
+  }
+
+  int Connect(int sockfd, const struct sockaddr* addr, socklen_t addrlen) {
+    if ((addr->sa_family == AF_INET && !ipv4_supported_) ||
+        (addr->sa_family == AF_INET6 && !ipv6_supported_)) {
+      errno = EAFNOSUPPORT;
+      return -1;
+    }
+    char* ip_addr_str;
+    grpc_resolved_address resolved_addr;
+    memcpy(&resolved_addr.addr, addr, addrlen);
+    resolved_addr.len = addrlen;
+    grpc_sockaddr_to_string(&ip_addr_str, &resolved_addr,
+                            false /* normalize */);
+    auto it = dest_addr_to_src_addr_.find(ip_addr_str);
+    if (it == dest_addr_to_src_addr_.end()) {
+      gpr_log(GPR_DEBUG, "can't find |%s| in dest to src map", ip_addr_str);
+      errno = ENETUNREACH;
+      return -1;
+    }
+    fd_to_getsockname_return_vals_.insert(
+        std::pair<int, TestAddress>(sockfd, it->second));
+    return 0;
+  }
+
+  int GetSockName(int sockfd, struct sockaddr* addr, socklen_t* addrlen) {
+    auto it = fd_to_getsockname_return_vals_.find(sockfd);
+    EXPECT_TRUE(it != fd_to_getsockname_return_vals_.end());
+    grpc_resolved_address resolved_addr =
+        TestAddressToGrpcResolvedAddress(it->second);
+    memcpy(addr, &resolved_addr.addr, resolved_addr.len);
```

Comment:
You need to check `*addrlen` before writing to this buffer.  The current implementation sets `*addrlen=0`, so the tests *should* be failing.

Reply:
ah this was not honoring description of that parameter, fixed